### ISSUE TITLE
Deprecation Guideline Proposal

### ISSUE THREAD

**Issue by BSFishy** [opensearch-project/oui#647]
Right now, we don't have guidelines on deprecation in OUI. It is important that we establish them so that we have a consistent and streamlined approach in the future.

There are 2 main things we might be deprecating: components and component props. Here is my proposal for guidance on deprecating both.

These are just drafts on what this guidance should be, feel free to drop any questions, comments, or suggestions.

### Components

#606 establishes a phenomenal precedent on deprecating components. Namely, doing these things:

 1. Wrap the component in `@opensearch-project/oui/utils/deprecated` to log a message when a deprecated component is being used (helpful for consumers to know when they're using deprecated components)
 2. Add a doc comment on the component saying the component is deprecated (helpful for consumers to see an IDE deprecation message when writing code)

The first point helps consumers refactor code that has already been written using deprecated components and the second point helps prevent consumers from writing new code that uses deprecated components.

### Component Props

This is new water, so it may go through a few iterations before we land on guidance that we stick with. I think there are a few steps that _should_ be taken, then a few that we may or may not add to the guidance:

 1. Add a doc comment to the prop saying the prop is deprecated (helpful for consumers to see an IDE deprecation message when writing code)
 2. Reformat code to use both old and new prop, if there is a new replacement prop (helpful for consumers for compatibility)
 3. Add a console log message whenever the deprecated prop is used (helpful for consumers to know when they're using deprecated props)

**Comment by KrooshalUX** [comment]
@ashwin-pc you are correct that in the *ideal* situation, we would deprecate something and have the deprecation notice also available on the docs site in sync with the deprecation notice within source code. 100% aligned, lets be sure that OUI docs site is called out explicitly in all of these guidelines. 

Just to give some context as to why we _didn't_ follow that in the case of `OuiLoadingElastic` and `OuiLoadingKibana` it was simply for compliance reasons and to get out ahead of a flood of "there are references to Kibana and Elastic on the OUI docs site" issues filed by viewers and to not block the go-live of the documentation.

**Comment by KrooshalUX** [comment]
Some additional thoughts about buckets outside of "component" and "component props": 

- Icons - we are about to deprecate a ton more of these after #249, and we also already hid a number of icons from the docs site launch but have not yet "deprecated" the icons from a code perspective. This was also to achieve compliance (ex: removing the elastic logo from the page so it is not referenced by designers and engineers) 
- Themes? 
- Sass variables?
- Utilities - do these count as "components" or are they treated differently?

**Comment by BSFishy** [comment]
> 1. Will adding the deprecated JSDoc annotation show up in the docsite? Since that is what users will refer to when they build using OUI?
> 2. The same goes for components. In the case of [[CCI] Add deprecation comments to OuiLoadingElastic and OuiLoadingKibana #606](https://github.com/opensearch-project/oui/pull/606) we dont want it to exist on the docsite for obvious reasons, but that will not be true in future. In future, even component deprecations need to show up in the docsite since it has to be an accurate reflection of the state of the library's public contract.

1. The JSDoc won't show up on the docsite, added another point to update the docsite
2. Same

**Comment by BSFishy** [comment]
> Some additional thoughts about buckets outside of "component" and "component props":
> 
> * Icons - we are about to deprecate a ton more of these after [(Eng Research) Investiage the use of Logos & App Icons  #249](https://github.com/opensearch-project/oui/issues/249), and we also already hid a number of icons from the docs site launch but have not yet "deprecated" the icons from a code perspective. This was also to achieve compliance (ex: removing the elastic logo from the page so it is not referenced by designers and engineers)
> * Themes?
> * Sass variables?
> * Utilities - do these count as "components" or are they treated differently?

 * Updated to add icons. We'll need to implement what I'm proposing, but it should be a super simple system once it is implemented.
 * I want to deprecate our entire current theming system. Not sure when it will be, but I don't think we'll need guidance on something that is already deprecated. If the new theming system works out the way I want it to, deprecating themes will be similar to deprecating utilities.
 * In the new theming system, we're completely moving away from Sass variables. Similar to the point above, I don't see the point in coming up with guidance on deprecating something that's already deprecated.
 * Updated the issue to add utilities.

**Comment by ashwin-pc** [comment]
@KrooshalUX yeah i understand why we had to do it this way this time, but was just calling out that thats the exception and not the norm :)

**Comment by ashwin-pc** [comment]
I would also like to know some of the options you considered for this proposal. Deprecation strategies for UI systems is an already solved problem. Many mature UI systems already do this all the time. Are we following the same pattern? If we arent, why are we deviating?

**Comment by SergeyMyssak** [comment]
I think this proposal is very well written. As an addition to this proposal, I assume that we need to create a separate page that describes the steps to migrate from one version to another. Here are examples of how this has been done in other popular ui libraries:
1) material-ui (85.9k stars) https://v5-0-6.mui.com/guides/migration-v4/#introduction
2) ant-design (85.5k stars) https://ant.design/docs/react/migration-v5
3) react-bootstrap (21.7k stars) - https://react-bootstrap.github.io/migrating/

It would be good to summarise all this and create action items to implement this deprecation guideline.

**Comment by anirudha** [comment]
Closing — this discussion/proposal has been inactive for over a year with no traction. If there's renewed interest, a fresh issue with updated context would be more productive.
